src/librustc_mir_build/hair/pattern/_match.rs

```diff
@@ -27,8 +29,52 @@
 /// pattern to those that have come before it doesn't increase the number of values
 /// we're matching).
 ///
+/// # Core concept
+///
+/// The idea that powers everything that is done in this file is the following: a value is made
+/// from a constructor applied to some fields. Examples of constructors are `Some`, `None`, `(,)`
+/// (the 2-tuple constructor), `Foo {..}` (the constructor for a struct `Foo`), and `2` (the
+/// constructor for the number `2`). Fields are just a (possibly empty) list of values.
+///
+/// Some of the constructors listed above might feel weird: `None` and `2` don't take any
+/// arguments. This is part of what makes constructors so general: we will consider plain values
+/// like numbers and string literals to be constructors that take no arguments, also called "0-ary
+/// constructors"; they are the simplest case of constructors. This allows us to see any value as
+/// made up from a tree of constructors, each having a given number of children. For example:
+/// `(None, Ok(0))` is made from 4 different constructors.
+///
+/// This idea can be extended to patterns: a pattern captures a set of possible values, and we can
+/// describe this set using constructors. For example, `Err(_)` captures all values of the type
+/// `Result<T, E>` that start with the `Err` constructor (for some choice of `T` and `E`). The
+/// wildcard `_` captures all values of the given type starting with any of the constructors for
+/// that type.
+///
+/// We use this to compute whether different patterns might capture a same value. Do the patterns
+/// `Ok("foo")` and `Err(_)` capture a common value? The answer is no, because the first pattern
+/// captures only values starting with the `Ok` constructor and the second only values starting
+/// with the `Err` constructor. Do the patterns `Some(42)` and `Some(1..10)` intersect? They might,
+/// since they both capture values starting with `Some`. To be certain, we need to dig under the
+/// `Some` constructor and continue asking the question. This is the main idea behind the
+/// exhaustiveness algorithm: by looking at patterns constructor-by-constructor, we can efficiently
+/// figure out if some new pattern might capture a value that hadn't been captured by previous
+/// patterns.
+///
+/// Constructors are represented by the `Constructor` enum, and its fields by the `Fields` enum.
+/// Most of the complexity of this file resides in transforming between patterns and
+/// (`Constructor`, `Fields`) pairs, handling all the special cases correctly.
+///
+/// Caveat: this constructors/fields distinction doesn't quite cover every Rust value. For example
+/// a value of type `Rc<u64>` doesn't fit this idea very well, nor do function pointers and various
+/// other things. However, the idea covers everything that can be pattern-matched, and this is all
+/// we need for exhaustiveness checking.
```

Comment:
Except that function pointers (and raw pointers) can be pattern-matched...
```rust
fn foo() {}

fn bar(x: fn()) -> bool {
    match x {
        foo => true,
        _ => false,
    }
}
```

Reply:
Yeahhhh, I wasn't sure whether to include that. I felt like that is the kind of detail I can swipe under the rug in a high-level overview, like uninhabited types. Would you prefer I mention this?

Reply:
I guess I could just remove this example actually. Can I still update the PR now?

Reply:
@Nadrieril You can update it, but it will need to be reapproved to enter the queue (that's probably ok, because I doubt it would make it to the head of the queue today anyway).

Reply:
When I wrote the comment, the text explicitly said that function pointers cannot be pattern matched. I am okay with sweeping details under the rug but that was just false information. ;)

Reply:
Ah yeah, I understand how you were reading it now. I haven't meant to iloly that they weren't matchable at all